### Project Introduction

Phishing is one of the most common cyberattacks used to steal sensitive information such as usernames, passwords, banking credentials, and personal data. Attackers create fraudulent websites that closely resemble legitimate websites to deceive users into revealing confidential information.

Traditional phishing detection techniques struggle to identify newly created or previously unseen phishing attacks.

Machine Learning provides a more adaptive approach by learning patterns from historical phishing and legitimate URLs. Instead of relying solely on predefined rules, machine learning algorithms analyse various URL, domain, DNS, SSL, and WHOIS-related characteristics to classify a website as either legitimate or phishing.

### Problem Statement

Phishing attacks continue to be one of the most significant cybersecurity threats, targeting individuals and organisations by impersonating legitimate websites to steal sensitive information. Detecting phishing websites manually is difficult due to the large number of URLs generated daily and the constantly evolving techniques used by attackers.

The objective of this project is to develop a machine learning model capable of automatically classifying a URL as either **Legitimate** or **Phishing** based on engineered URL, DNS, SSL, and domain-related features.

This is a **supervised binary classification** problem, where the model learns from labelled examples and predicts one of two possible classes:

- **Legitimate (0)** – A genuine and trusted website.
- **Phishing (1)** – A fraudulent website designed to deceive users.

The primary objective is to maximise the model's ability to correctly identify phishing websites while minimising false predictions. Multiple machine learning algorithms will be trained, evaluated, and compared using appropriate classification metrics to identify the best-performing model for phishing URL detection.

---

### 1. Dataset Information

#### Original Dataset

The foundation of this project is the **PhiUSIIL Phishing URL Dataset**, obtained from the UCI Machine Learning Repository. The original dataset contains labelled URLs together with a large number of pre-engineered features designed for phishing detection.

Rather than using the provided engineered features directly, only the **URL** and **Label** columns were retained. This decision was made to develop a completely custom feature engineering pipeline capable of extracting meaningful characteristics directly from raw URLs.

By starting with only the raw URL and its corresponding label, the project demonstrates the complete workflow of a real-world phishing detection system, where feature extraction is performed before machine learning model training. This approach provides greater flexibility, transparency, and a deeper understanding of the phishing detection process.

#### What was measured
What are the effective minimal sets of features that can be utilized in predicting phishing?

#### Why it was collected
With the rise in cybercrime, phishing remains a significant concern as it targets individuals with fake websites, causing victims to disclose their private information. The effective implementation of phishing detection relies on cost efficiency, with the increased feature extraction factor contributing to these costs. This advances ongoing efforts to improve the accuracy and efficiency of phishing detection mechanisms
and enhance cybersecurity defenses against malicious activities. This aims to explore various methods for detecting phishing emails using machine learning techniques by systematically evaluating and comparing different models.

#### How the data was obtained

Unlike many phishing detection projects that directly use pre-engineered features, this project generates a completely new machine learning dataset through a custom feature extraction pipeline.

The original PhiUSIIL dataset was used only as a source of labelled URLs. During preprocessing, only the **URL** and **Label** columns were retained, while all existing engineered features were discarded. For each URL, multiple custom Python modules automatically extracted lexical, domain, SSL, WHOIS, DNS, and internet-based characteristics.

The dataset generation pipeline consists of three specialised feature extraction modules:

- **Basic Features** – Extracts lexical properties directly from the URL, such as URL length, number of dots, special characters, path complexity, and HTTPS usage.
- **Domain Features** – Extracts domain-related characteristics including subdomains, suspicious keywords, brand impersonation indicators, top-level domain analysis, URL entropy, and redirect masking.
- **Internet Features** – Collects network-based information including SSL certificates, WHOIS records, domain age, DNS information, HTTP responses, webpage structure, HTML forms, password fields, and external hyperlink statistics.

The extracted features from each module are merged together and combined with the original class label to generate a new dataset named **Phishing_Features.csv**. This generated dataset forms the foundation for all subsequent stages of the project, including exploratory data analysis, feature selection, machine learning model development, and performance evaluation.


**What do the instances in this dataset represent?**
URLs and their corresponding webpages

**Has Missing Values?**
Yes


### Project Structure

To improve maintainability, readability, and scalability, the phishing URL detection system was organised into multiple Python modules, with each module responsible for a specific stage of the feature extraction process. This modular architecture follows good software engineering practices by separating feature extraction, dataset generation, and machine learning into independent components.

```
Phishing_URL_Detector/
│
├── Phishing_URL_Dataset.csv      # Original dataset
├── extraction.py                 # Main feature extraction pipeline
├── basic_features.py             # Lexical feature extraction
├── dmn_features.py               # Domain feature extraction
├── internet_features.py          # SSL, WHOIS, DNS and HTTP features
├── Phishing_Features.csv         # Generated feature dataset
├── phishing_url_detector.ipynb   # Machine learning notebook
└── README.md                     # Project documentation
```

Each module has a clearly defined responsibility. The original dataset serves as the source of labelled URLs, the feature extraction modules generate meaningful characteristics from each URL, and the generated dataset is subsequently used to train and evaluate multiple machine learning models.

### Feature Categories

The custom feature extraction pipeline generates features from multiple sources to capture different characteristics of phishing websites. Rather than relying on a single type of information, the extracted features are grouped into logical categories, providing a comprehensive representation of each URL.

| Feature Category | Description | Example Features |
|------------------|-------------|------------------|
| **Lexical Features** | Features extracted directly from the URL string without internet access. | URL length, domain length, path length, number of dots, digits, special characters, HTTPS usage, IP address detection. |
| **Domain Features** | Features describing the structure and characteristics of the domain name. | Subdomain count, suspicious keywords, brand spoofing, URL entropy, top-level domain analysis, redirect masking. |
| **Internet Features** | Features requiring communication with external services. | SSL certificate information, WHOIS records, domain age, DNS records, HTTP status codes. |
| **Webpage Features** | Features extracted after analysing the webpage content. | HTML forms, password fields, suspicious page titles, external hyperlink ratio. |

By combining lexical, domain, internet, and webpage-based features, the generated dataset provides a comprehensive representation of each URL, enabling machine learning models to distinguish between legitimate and phishing websites more effectively.

---

### Understanding URL Structure

Before extracting features from a URL, it is important to understand its structure. A **URL (Uniform Resource Locator)** is the address used to locate resources on the Internet. Every URL consists of multiple components, and many of the features used in phishing detection are derived directly from these components.

For example, features such as the URL length, domain name, subdomains, path length, number of query parameters, and the protocol (HTTP or HTTPS) are all obtained by analysing different parts of a URL.

Understanding these components will make it easier to understand the feature engineering process later in this notebook.

---

### Example URL

```
https://login.secure-paypal.com.account-verification.xyz:8080/security/update/login.php?id=10293&session=abc123#section2
```

Let's break this URL into its individual components.

---

####  Scheme (Protocol)

```
https://
```

The **scheme**, also known as the **protocol**, specifies how communication takes place between the client (your browser) and the web server.

Common protocols include:

| Protocol | Description |
|----------|-------------|
| HTTP | HyperText Transfer Protocol (not encrypted) |
| HTTPS | HyperText Transfer Protocol Secure (encrypted using SSL/TLS) |
| FTP | File Transfer Protocol |

Although HTTPS encrypts communication, attackers can also obtain SSL certificates. Therefore, HTTPS alone does **not** guarantee that a website is legitimate.

---

####  Host

```
login.secure-paypal.com.account-verification.xyz
```

The **host** identifies the server that hosts the website.

It consists of:

- Registered Domain
- Subdomains

##### Registered Domain

```
account-verification.xyz
```

The registered domain is the domain that has been officially registered.

It consists of:

- **Domain Name:** `account-verification`
- **Top-Level Domain (TLD):** `xyz`

Common Top-Level Domains include:

- `.com`
- `.org`
- `.net`
- `.edu`
- `.gov`

---

##### Subdomains

```
login.secure-paypal.com
```

Everything before the registered domain is considered the **subdomain**.

Subdomains are often used to organise different sections of a website.

For example:

```
mail.google.com
```

- Subdomain: `mail`
- Registered Domain: `google.com`

Phishing websites frequently create long and misleading subdomains to trick users into believing they are visiting a trusted website.

For example:

```
login.paypal.com.fake-domain.xyz
```

Although the URL appears to contain **paypal.com**, the actual registered domain is **fake-domain.xyz**.

---

### Port (Optional)

```
:8080
```

The **port number** specifies which network service on the server should receive the request.

Common ports include:

| Port | Service |
|------|----------|
| 80 | HTTP |
| 443 | HTTPS |
| 21 | FTP |
| 22 | SSH |
| 8080 | Alternative HTTP |

If no port is specified, the browser automatically uses the default port associated with the protocol.

---

### Path

```
/security/update/login.php
```

The **path** specifies the location of the requested resource on the web server.

It may point to:

- A webpage
- An image
- A file
- A script
- An API endpoint

Long or deeply nested paths may sometimes indicate suspicious behaviour.

---

### Query Parameters

```
?id=10293&session=abc123
```

Query parameters provide additional information to the server.

Each parameter consists of a **key-value pair**.

For this example:

| Parameter | Value |
|-----------|-------|
| id | 10293 |
| session | abc123 |

Multiple parameters are separated using the `&` character.

Large numbers of query parameters or unusually long query strings may indicate suspicious URLs.

---

### Fragment (Optional)

```
#section2
```

A **fragment** identifies a specific section within a webpage.

Fragments are processed by the browser and are generally **not sent to the web server**.

---

### Complete URL Breakdown

```
https://login.secure-paypal.com.account-verification.xyz:8080/security/update/login.php?id=10293&session=abc123#section2

│
├── Scheme (Protocol)
│   https
│
├── Host
│   ├── Subdomains
│   │   login.secure-paypal.com
│   │
│   ├── Registered Domain
│   │   account-verification.xyz
│   │
│   └── Top-Level Domain
│       xyz
│
├── Port
│   8080
│
├── Path
│   /security/update/login.php
│
├── Query Parameters
│   id=10293
│   session=abc123
│
└── Fragment
    section2
```

---

### Why is Understanding URL Structure Important?

Many of the features extracted in this project are derived directly from different components of a URL. Understanding the structure of a URL makes it easier to understand how each feature is generated and why it may help distinguish legitimate websites from phishing websites.

| URL Component | Example Features |
|--------------|------------------|
| Entire URL | URL length, number of digits, number of special characters |
| Scheme | Uses HTTPS |
| Registered Domain | Domain length, suspicious keywords, Top-Level Domain |
| Subdomains | Number of subdomains |
| Path | Path length, file extension |
| Query Parameters | Number of parameters, query length |

---

### Loading the Dataset

Before we can do anything with machine learning, we need to load our data, and one of the most common formats for storing data is CSV, which stands for *Comma-Separated Values*. A CSV file is simply a plain text file where each row represents a record (like a person, a review, or a product), and each value is separated by a comma.

We will load CSV files into a program.
 
Using the `pandas` library and its `read_csv()` function (the most flexible and beginner-friendly)

It has its own strengths:

*pandas* is the most popular option because it makes working with tables of data very easy. It automatically handles many of the common problems you might run into and gives you a powerful tool called a *DataFrame*, a table that you can search, filter, and analyse with just a few lines of code.

Few things to check when loading a CSV file:

- *Header row*: Does the first line of your file contain column names (like "Name", "Age", or "Review")? If so, most tools can detect and use it automatically. If not, you'll need to tell the tool what the column names should be.

- *Comments*: Some CSV files include comment lines that start with a symbol like `#`. These lines are meant for humans, not the program, so you may need to tell your loading function to skip them.

- *Delimiter*: Although CSV stands for "comma-separated", sometimes other characters are used instead, like tabs (`\t`) or semicolons (`;`). Make sure you specify the correct one so the data loads properly.

---

In [16]:
# Install required libraries
%pip install pandas


Note: you may need to restart the kernel to use updated packages.


In [17]:
import pandas as pd

# Load the Phishing Dataset
filename = "Phishing_Features.csv"

# Read the dataset while preserving valid category values
# such as "None". This prevents pandas from automatically
# converting them into missing values (NaN).

df = pd.read_csv(
    filename,
    keep_default_na=False
)

### Checking the Dataset

After loading a dataset, it is good practice to inspect the first few rows before performing any analysis. This helps confirm that the dataset has been loaded correctly and provides a quick overview of its structure.

The head() function displays the first five rows of a DataFrame by default. It is commonly used to verify that the column names have been read correctly and that the values appear under the correct features.

We will now display the first five rows of the dataset to confirm that the data has been loaded successfully.

---

In [18]:
# Display the First Five Rows
df.head()

,URL,url_length,domain_length,path_length,query_length,param_count,has_equal,has_ampersand,query_ratio,dots,...,domain_age_days,new_domain,dns_exists,dns_ip_count,http_status,title_suspicious,form_count,has_password_field,external_link_ratio,label
0,https://www.southbankmosaics.com,32,24,0,0,0,0,0,0.0,2,...,2353,0,1,1,200,0,1,0,0.027777777777777776,1
1,https://www.uni-mainz.de,24,16,0,0,0,0,0,0.0,2,...,None,None,1,1,200,0,0,0,1.0,1
2,https://www.voicefmradio.co.uk,30,22,0,0,0,0,0,0.0,3,...,5931,0,1,4,200,0,0,0,0.22115384615384615,1
3,https://www.sfnmjournal.com,27,19,0,0,0,0,0,0.0,2,...,7218,0,1,2,403,None,None,None,None,1
4,https://www.rewildingargentina.org,34,26,0,0,0,0,0,0.0,2,...,2553,0,1,2,200,0,1,0,0.7647058823529411,1


### Understanding the Output

The output displays the first five rows of the generated phishing feature dataset.

Each row represents one URL, while each column represents an extracted feature or the corresponding class label.

At this stage, we are simply verifying that the dataset has been loaded correctly. We are not yet analysing the feature values themselves.

---

### Dataset Dimensions

Before performing exploratory data analysis or building machine learning models, it is useful to determine the size of the dataset.

The size of a dataset is described by its **dimensions**, which consist of:

- **Rows** – The number of observations (URLs).
- **Columns** – The number of variables, including the extracted features and the target label.

Pandas provides the `shape` attribute to return the dimensions of a DataFrame as a tuple in the form:

```
(number_of_rows, number_of_columns)
```

Knowing the dimensions of the dataset helps verify that the dataset has been loaded completely and provides an overview of its size before further analysis.

---

In [19]:
# How big is our data
print(df.shape)

(235795, 58)


### Understanding the Output

The output is displayed as a tuple in the following format:

```
(number_of_rows, number_of_columns)
```

For example:

```
(235795, 58)
```

This means:

- The dataset contains **235,795 URLs (rows)**.
- The dataset contains **58 columns**, including the engineered features and the target label.

Each row represents a single URL processed by the feature extraction pipeline, while each column represents either an extracted feature or the corresponding class label.

---

### Exploring more details of the data

Beyond checking the first few and last few rows, pandas offers convenient methods to quickly summarise your dataset. For instance, if you want to view column data types and any non-null counts, you can use data.info(). This can be useful to spot missing values or confirm that columns are numeric:

---

In [20]:
# Checking data info
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 235795 entries, 0 to 235794
Data columns (total 58 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   URL                  235795 non-null  str    
 1   url_length           235795 non-null  int64  
 2   domain_length        235795 non-null  int64  
 3   path_length          235795 non-null  int64  
 4   query_length         235795 non-null  int64  
 5   param_count          235795 non-null  int64  
 6   has_equal            235795 non-null  int64  
 7   has_ampersand        235795 non-null  int64  
 8   query_ratio          235795 non-null  float64
 9   dots                 235795 non-null  int64  
 10  hyphens              235795 non-null  int64  
 11  underscores          235795 non-null  int64  
 12  letters_ratio        235795 non-null  float64
 13  digit_ratio          235795 non-null  float64
 14  special_ratio        235795 non-null  float64
 15  https                235795 

From `data.info()`, we get a quick overview of the dataset's structure. It tells us:

- How many rows are in the dataset  
- Which columns are present  
- How many non-null (non-missing) values are in each column  
- What data type each column uses (such as `int64` for whole numbers, `float64` for decimal numbers, or `object` for text)

This information is especially helpful for spotting issues early. For example:
- If a column has fewer non-null values than the total number of rows, it likely contains *missing data*.
- If a column that should contain numbers is listed as `object`, the values might be stored as text, possibly due to formatting issues (like commas or symbols).

Checking the results of `data.info()`, allows us to make sure the dataset is properly formatted and ready for analysis or model training.

---

###  Analysing , visualising and Transforming tabular data

### Introduction  

We will explore how to analyse and preprocess a numeric/tabular dataset to gain meaningful insights before applying machine learning models. Understanding your data is a crucial first step in any data science workflow, as it influences how well your models perform and generalise to unseen data.

We will cover three key aspects of data exploration and preparation:  

- *Describing the dataset*: we will generate statistical summaries, such as measures of central tendency (mean, median) and dispersion (variance, standard deviation), to understand the distribution of each feature. Additionally, we will examine missing values, outliers, and correlations between variables to identify patterns and potential preprocessing steps.  

- *Visualising the dataset*: data visualisation helps us uncover hidden relationships, trends, and anomalies. We will use histograms, box plots, scatter plots, and correlation heatmaps to visually inspect distributions, detect skewness, and understand feature interactions.    
**Visualisations will be introduced throughout the notebook whenever they help explain the data. This approach ensures that each visualisation has a clear purpose rather than being included unnecessarily.**

This process is essential for making informed decisions in machine learning, as the quality of your input data directly impacts model accuracy and reliability.

###  Describing the Dataset

#### Statistical summary

Descriptive statistics help us understand how each part of a dataset is spread out. When we run `.describe()` on a pandas DataFrame in Python, we get a helpful summary that includes:

- *Count*:  </br>
Tells us how many values are present (not missing) for each column. This helps us spot any gaps in the data.

- *Mean*:  </br>
The average value. This gives us a rough idea of what's "typical" for each column.

- *Standard Deviation*:</br>
Shows how much the values vary. A high number means the data is more spread out.

- *Minimum Value*:  </br>
The smallest number in the column. This can help us find any errors or outliers.

- *25th Percentile (Q1)*:  </br>
A quarter of the values fall below this point. It gives us a sense of the lower end of the data.

- *50th Percentile (Median/Q2)*:  </br>
The middle value when the data is sorted. This shows us the centre of the data.

- *75th Percentile (Q3)*:  </br>
Three-quarters of the data fall below this point. It helps us understand the upper range.

- *Maximum Value*:  </br>
The largest value in the column. This can highlight extreme values or mistakes.

These summaries make it easier to spot missing information, unusual results, or parts of the data that might need special attention, like features with a very wide range of values:

---

In [21]:
# Summary Statistics 
df.describe()

,url_length,domain_length,path_length,query_length,param_count,has_equal,has_ampersand,query_ratio,dots,hyphens,...,deep_subdomain,very_deep_subdomain,php_page,long_path,path_depth,url_encoded,entropy,redirect_masking,has_at_misuse,label
count,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000,...,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000
mean,35.366954,21.470396,3.735325,2.212977,0.060841,0.023270,0.011951,0.011065,2.257143,0.350249,...,0.022129,0.007795,0.022414,0.025217,0.419746,0.003762,3.960896,0.006391,0.006349,0.571895
std,41.214346,9.150793,23.501818,24.880892,0.555041,0.150761,0.108666,0.080187,0.922678,1.521763,...,0.147105,0.087944,0.148025,0.156783,0.970301,0.061218,0.315386,0.079689,0.079426,0.494805
min,14.000000,4.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,2.308993,0.000000,0.000000,0.000000
25%,24.000000,16.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,2.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.763342,0.000000,0.000000,0.000000
50%,28.000000,20.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,2.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.929214,0.000000,0.000000,1.000000
75%,35.000000,24.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,2.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,1.000000,0.000000,4.100817,0.000000,0.000000,1.000000
max,6097.000000,110.000000,4248.000000,3391.000000,51.000000,1.000000,1.000000,0.976108,99.000000,286.000000,...,1.000000,1.000000,1.000000,1.000000,33.000000,1.000000,5.335261,1.000000,1.000000,1.000000


To explore the categorical features in the Diabetes Hospital Readmission dataset, we use the `describe()` function with the parameter `include=["object"]`. This tells pandas to generate summary statistics only for columns stored as the **object** data type, which in this dataset represents the categorical features.

The output of `describe()` is displayed using **`.T` (transpose)**, which swaps the rows and columns of the table. This does not change the data itself; it simply makes the output easier to read by displaying one feature per row.


In [22]:
# Summary Statistics for Categorical Features
df.describe(include=["object"]).T

/var/folders/6w/6jhfgq8s7hb_9__z7h3g3tcm0000gn/T/ipykernel_34017/4028559741.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include=["object"]).T


,count,unique,top,freq
URL,235795,235370,https://att-104164.weeblysite.com/,2
subdomain,235795,39657,www,161291
domain,235795,169415,web,5757
tld,235795,1402,com,112565
registered_domain,235795,175509,web.app,5754
has_ssl,235795,3,1,161119
ssl_issuer_known,235795,1,None,235795
domain_age_days,235795,12477,None,96391
new_domain,235795,3,0,137518
dns_exists,235795,2,1,187561


### Understanding the Output

Unlike numerical features, categorical features are summarised using different statistics because they represent labels or categories rather than continuous numerical values.

The output includes:

- **Count** – The number of non-null values in the feature.
- **Unique** – The number of distinct categories present in the feature.
- **Top** – The most frequently occurring category (mode).
- **Freq** – The number of times the most common category appears.

These statistics provide a quick overview of the categorical features and help us understand the distribution of different categories before performing data preprocessing or feature engineering.

It is important to interpret these statistics carefully. During feature extraction, unavailable values were intentionally stored as the literal string `"None"` rather than being converted to `NaN`. Furthermore, the dataset was loaded using `keep_default_na=False`, ensuring that `"None"` remains a valid category instead of being treated as a missing value.

As a result, placeholder values such as `"None"` are included when calculating the **Count**, **Unique**, **Top** and **Freq** statistics. In some features, `"None"` may even appear as the most frequently occurring category.

This design provides greater flexibility during preprocessing. Rather than immediately treating unavailable values as missing data, each feature can be analysed individually. For example, if a feature contains a very high proportion of `"None"` values, it may be removed. Conversely, if only a small proportion of values are `"None"`, they can be retained as a separate category because the absence of information itself may provide useful information for distinguishing phishing websites from legitimate ones.

A dedicated missing value analysis will therefore be performed later in the notebook to determine the most appropriate handling strategy for each feature.

---